Adding Whats New modal to the block editor

donlair opened the pull request:
Adding a new option to the edit post menu in Gutenberg for WPcom (pau2Xa-2eS-p2 for context).

This PR adds a What's New Guide in the WPcom block editor that highlights a predefined list of new features.

The guide is accessible via the sidebar menu circled in this screenshot:
<img width="1516" alt="Screen Shot 2021-01-21 at 10 08 28 AM" src="https://user-images.githubusercontent.com/35781181/105371307-7a6aa000-5bd2-11eb-874c-ed6a8177930a.png">

Clicking that icon pops the sidebar menu. You'll see a new "What's New" option in the "Tools" submenu. Clicking that link pops the What's New guide, which looks like this:
<img width="1200" alt="Screen Shot 2021-01-21 at 10 11 57 AM" src="https://user-images.githubusercontent.com/35781181/105371447-9ff7a980-5bd2-11eb-8412-de26beeb1569.png">

#### Testing instructions

* Checkout this PR and sync the editor toolkit plugin to your sandbox.
* Sandbox one of your test sites.
* Load the post editor and navigate the menus as shown above.
* Load the What's New guide, read the cards, and click the links to confirm they work as expected.

Fixes #

donlair committed:
Attempting a skeleton implementation for Whats New

donlair committed:
Tweaks

noahtallen commented:
In calypso, we use `yarn` instead of `npm`, so make sure to delete the `package-lock.json` files that got added. :) Then, install everything with `yarn install`, and run local build commands with yarn as well: `cd apps/editing-toolkit && yarn dev --sync`

noahtallen commented on apps/editing-toolkit/package.json:
```diff
@@ -117,7 +120,7 @@
 		"@wordpress/dom-ready": "*",
 		"@wordpress/e2e-test-utils": "^4.11.2",
-		"@wordpress/edit-post": "*",
+		"@wordpress/edit-post": "^3.26.1",
```
This change should not be needed

noahtallen commented on apps/editing-toolkit/package.json:
```diff
@@ -152,6 +155,7 @@
 	},
 	"devDependencies": {
 		"@babel/preset-typescript": "^7.12.1",
+		"@types/wordpress__edit-post": "^3.5.4",
```
I don't know if it makes sense to add this one either -- I'd probably hold off on modifying wordpress-related dependencies right now because of an ongoing effort to update them :)

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/menu/index.js:
```diff
@@ -0,0 +1,19 @@
+ * WordPress dependencies
+ */
+import { registerPlugin } from '@wordpress/plugins';
+import { ToolsMoreMenuGroup } from '@wordpress/edit-post';
```
I think the reason this doesn't work is that `ToolsMoreMenuGroup` is not exported from `@wordpress/edit-post`.

Instead, we need to reference the Fill manually. We need to import the `Fill` component from `@wordpress/components`

```suggestion
import { Fill } from '@wordpress/components';
```

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/whats-new.php:
```diff
@@ -0,0 +1,69 @@
+	 * Whats_New constructor.
+	 */
+	public function __construct() {
+		add_action( 'enqueue_whats_new_assets', array( $this, 'enqueue_script_and_style' ), 100 );
```
I think this doesn't quite work as you expect: `add_action` in this case is defining an action called "enqueue_whats_new_assets". But that action is never called anywhere. For an action to do anything, you would have to write `do_action( 'enqueue_whats_new_assets' )`.

Basically, the problem is that you actually just want this function (`enqueue_script_and_style`) to run whenever a _different_ action runs. That action is called `enqueue_block_editor_assets`, and Gutenberg calls `do_action( 'enqueue_block_editor_assets' )` when it is loading: so if you hook into that action, your function here will be run when Gutenberg tells it to.

It just needs to be changed to:
```suggestion
		add_action( 'enqueue_block_editor_assets', array( $this, 'enqueue_script_and_style' ), 100 );
```

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/whats-new.php:
```diff
@@ -0,0 +1,69 @@
+
+		wp_set_script_translations( 'wpcom-whats-new', 'full-site-editing' );
+
+		$style_path = 'dist/whats-new' . ( is_rtl() ? '.rtl' : '' ) . '.css';
+		wp_enqueue_style(
+			'wpcom-whats-new',
+			plugins_url( $style_path, __FILE__ ),
+			array(),
+			filemtime( plugin_dir_path( __FILE__ ) . $style_path )
+		);
```
this enqueue call will fail at the moment because there is no CSS file yet -- so you'll need to comment it out to test.

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/menu/index.js:
```diff
@@ -0,0 +1,19 @@
+import { ToolsMoreMenuGroup } from '@wordpress/edit-post';
+
+const WhatsNew = () => (
+	<ToolsMoreMenuGroup>
+		<p>What's New copy 2</p>
+	</ToolsMoreMenuGroup>
```
Then, update this component to use the fill manually, and give it the right name:

```suggestion
	<Fill name="ToolsMoreMenuGroup">
		<p>What's New copy 2</p>
	</Fill>
```

noahtallen reviewed (dismissed):
Yeah, make sure to remove both `package-lock.json` files (and I think also the changes to the `package.json` file), and then to install the dependencies by running the `yarn` command from the root of the wp-calypso repository.

With the changes I suggested in the PR, it starts working:

<img width="1792" alt="Screen Shot 2021-01-08 at 3 19 28 PM" src="https://user-images.githubusercontent.com/6265975/104074679-94b18080-51c5-11eb-9f23-ee6099175463.png">

At this point, I would start looking at the components available in `@wordpress/components` in order to build out the item here. For example, use the `MenuItem` component to wrap the text here. It will have consistent style with the other items, and will act as a button.

donlair committed:
Apply suggestions from code review

Co-authored-by: Noah Allen <noahtallen@gmail.com>

donlair committed:
Addressing review feedback.

donlair committed:
adding MenuItem component

donlair committed:
placeholder

a8ci18n commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new-card.js:
```diff
@@ -0,0 +1,210 @@
+				slideNumber={ cardIndex + 1 }
+			/>
+			<CardMedia>
+				<img alt={ __( "What's New", 'full-site-editing' ) } src={ imgSrc } />
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **26** times:
`translate( 'What's new' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/whats-new)

donlair committed:
Still experimenting

donlair committed:
Rebase and cleanup.

a8ci18n commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/menu.js:
```diff
@@ -0,0 +1,120 @@
+		<>
+			<Fill name="ToolsMoreMenuGroup">
+				<MenuItem onClick={ () => toggleWhatsNew() }>
+					{ __( "What's New", 'full-site-editing' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **26** times:
`translate( 'What's new' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/whats-new)

donlair committed:
First functioning commit

a8ci18n commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new.js:
```diff
@@ -0,0 +1,122 @@
+		<>
+			<Fill name="ToolsMoreMenuGroup">
+				<MenuItem onClick={ () => toggleWhatsNew() }>
+					{ __( "What's New", 'full-site-editing' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **26** times:
`translate( 'What's new' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/whats-new)

noahtallen replied:
We should apply this change -- I think we may also just capitalize the first letter in general for copy in Calypso.

donlair committed:
Fixing link implementation in card.

donlair committed:
Card updates

donlair committed:
layout tweaks.

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php:
```diff
@@ -340,3 +340,11 @@ function load_coming_soon() {
+/**
+ * What's New section of the Tools menu
+ */
+function load_whats_new() {
```
We should probably restrict this based on when it needs to be loaded. In this case, we probably want to make sure that it is only available when Gutenberg is loaded.

noahtallen commented:
We'll eventually need to put the images elsewhere. I'm not sure what the best practice is right now, but they cannot be committed to dotcom via the plugin using the current approach. I imagine we might want to serve images over the CDN, but I don't know much about best practices for media in this case.

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new.js:
```diff
@@ -0,0 +1,186 @@
+	imgSrc,
+} ) {
+	useEffect( () => {
+		recordTracksEvent( 'calypso_editor_whats_new_slide_view', {
```
`calypso_editor` might not be the most descriptive, since this can also be loaded in wp-admin views outside of calypso. (Unless there is already a tracks event for this :))

donlair committed:
Adding single page and drop and drop cards.

donlair committed:
Card and package updates.

donlair committed:
Swapping images to static images hosted outside the editing toolkit.

donlair committed:
Updating Tracks event name.

donlair commented:
> We'll eventually need to put the images elsewhere. I'm not sure what the best practice is right now, but they cannot be committed to dotcom via the plugin using the current approach. I imagine we might want to serve images over the CDN, but I don't know much about best practices for media in this case.

I moved the images to the s0 server.

> calypso_editor might not be the most descriptive, since this can also be loaded in wp-admin views outside of calypso. (Unless there is already a tracks event for this :))

Changed to block editor so it's not so specific.

donlair committed:
Updating php filename.

donlair committed:
Correcting filename reference.

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new.js:
```diff
@@ -0,0 +1,247 @@
+					className="whats-new"
+					contentLabel={ __( "What's New at WordPress.com", 'full-site-editing' ) }
+					finishButtonText={ __( 'Close', 'full-site-editing' ) }
+					onFinish={ () => toggleWhatsNew() }
```
```suggestion
					onFinish={ toggleWhatsNew }
```

Nit: you can just reference the function as a variable directly :)

noahtallen commented on apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new.js:
```diff
@@ -0,0 +1,247 @@
+	return [
+		{
+			imgSrc: blockPatternsImage,
+			heading: __( 'New Block Patterns', 'full-site-editing' ),
```
nit: the other items only capitalize the first letter, so maybe `New block patterns` would be more consistent?

noahtallen reviewed (commented):
This tests very well for me! Can we also do the following:

1. Rebase on `trunk` (makes it easier to test; I think some changes are missing)
2. Remove the unused image files. (At least, I think they are unused.)

Otherwise, this is looking very good :)

donlair committed:
Addressing review comments.

donlair committed:
Updating cards to incorporate design feedback.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5495450

Thank you @donlair for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5495450).
